CH09: whole program compiler driver

esc opened the pull request:
Next attempt at the whole program compiler.

For the program at:

https://gist.github.com/sklam/13c0646a4e6b2d401b731835629b1be4

this currently produces.

```
 💣 zsh» python wpc.py llm.py
{'softmax': SymbolInfo(name='softmax',
                       ast=<ast.FunctionDef object at 0x104e9ff10>,
                       calls=[<ast.Attribute object at 0x104e9fc50>,
                              <ast.Attribute object at 0x104e9fb10>,
                              <ast.Attribute object at 0x104e9f710>]),
 'scaled_dot_product_attention': SymbolInfo(name='scaled_dot_product_attention',
                                            ast=<ast.FunctionDef object at 0x104e9f310>,
                                            calls=[<ast.Attribute object at 0x104e9ed90>,
                                                   <ast.Attribute object at 0x104e9e550>,
                                                   <ast.Attribute object at 0x104e9dd10>,
                                                   <ast.Attribute object at 0x104e9ce50>,
                                                   <ast.Attribute object at 0x104e9c810>,
                                                   <ast.Name object at 0x104e9c450>,
                                                   <ast.Attribute object at 0x104e9bf50>,
                                                   <ast.Attribute object at 0x104e9bb90>]),
 'MultiHeadAttention.__init__': SymbolInfo(name='MultiHeadAttention.__init__',
                                           ast=<ast.FunctionDef object at 0x104e9b610>,
                                           calls=[]),
 'MultiHeadAttention.split_heads': SymbolInfo(name='MultiHeadAttention.split_heads',
                                              ast=<ast.FunctionDef object at 0x104e9ac10>,
                                              calls=[<ast.Attribute object at 0x104e99f90>,
                                                     <ast.Attribute object at 0x104e99a50>]),
 'MultiHeadAttention.combine_heads': SymbolInfo(name='MultiHeadAttention.combine_heads',
                                                ast=<ast.FunctionDef object at 0x104e99510>,
                                                calls=[<ast.Attribute object at 0x104e98750>]),
 'MultiHeadAttention.forward': SymbolInfo(name='MultiHeadAttention.forward',
                                          ast=<ast.FunctionDef object at 0x104e8fd90>,
                                          calls=[<ast.Attribute object at 0x104e8f790>,
                                                 <ast.Attribute object at 0x104e8f3d0>,
                                                 <ast.Attribute object at 0x104e8f010>,
                                                 <ast.Name object at 0x104e8d890>,
                                                 <ast.Attribute object at 0x104e8d550>]),
 'FeedForwardNetwork.__init__': SymbolInfo(name='FeedForwardNetwork.__init__',
                                           ast=<ast.FunctionDef object at 0x104e8c750>,
                                           calls=[<ast.Attribute object at 0x104e8dd90>,
                                                  <ast.Attribute object at 0x104e8e2d0>]),
 'FeedForwardNetwork.forward': SymbolInfo(name='FeedForwardNetwork.forward',
                                          ast=<ast.FunctionDef object at 0x104e8e690>,
                                          calls=[<ast.Attribute object at 0x104ea9290>,
                                                 <ast.Attribute object at 0x104ea8f10>]),
 'TransformerLayer.__init__': SymbolInfo(name='TransformerLayer.__init__',
                                         ast=<ast.FunctionDef object at 0x104ea8750>,
                                         calls=[<ast.Name object at 0x104ea9910>,
                                                <ast.Name object at 0x104ea9d50>]),
 'TransformerLayer.forward': SymbolInfo(name='TransformerLayer.forward',
                                        ast=<ast.FunctionDef object at 0x104eaa1d0>,
                                        calls=[<ast.Attribute object at 0x104eaa810>,
                                               <ast.Attribute object at 0x104eaad50>])}
```

esc commented:
Hi @sklam @seibert,

this chapter has now been converted to the jupyter format used by the other chapters. Also, we can now connect call graphs across class attributes using the provided type annotations. I have tested both the ipynb and also the html variants.

Also, this includes a modification of the `Makefile` such that no `*.ipynb` files are built for ancillary files, i.e. everything that doesn't start with `ch` or `demo`: https://github.com/numba/numba-prototypes/pull/16/commits/3037f3d4aa73e0ebf56fa5828282ed08eadcba4e

Technically, this does not yet "work" in the sense that an actual program can be compiled. The completion of that capability is still waiting on:

a) An interface to the new jit compiler that can take an `ast`. The symbol info objects of this driver include the raw `ast`, which can be handed off to ` jit_compile` function, for example.

b) A solution as to what to do with the `np.*` calls. We either need out own implementations, like Numpy re-written in `mcl` or a way to link into the existing Numpy implementation, for example, by way of PIXIE or similar.

Essentially, in order to make this a "working" compiler, some more blanks must be filled in, but the overall structure to analyse the source code and extract the information required to perform the compile of the functions in a python module is implemented.

sklam commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -0,0 +1,326 @@
+        if call_qname in self.classes:
+            # If the call ends with the current class name, we replace it with
+            # the constructor call, since this is the Python semantics.
+            call_qname = call_qname + ".__init__"
```
This is good enough for now.

In the future, this can be quite complicated; e.g. `__new__` and metaclass stuffs.

esc replied:
yes

sklam commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -0,0 +1,326 @@
+# ## About
+#
+# In this chapter we will focus on the development of a "Whole Program Compiler
+# Drive". Essentially this is a high level part of a compiler which ties
```
```suggestion
# Driver". Essentially this is a high level part of a compiler which ties
```

sklam commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -0,0 +1,326 @@
+
+# TODO how to exclude this from jupyter execution.
+to_graphviz(cgv)
+
```
One more thing, can you add a test file in the tests directory for this chapter?

The test can just be asserting for the correct output from the function dependency of the global_calls.

esc replied:
Encountered a problem here. The global call:

```
output = transformer_layer.forward(input_data)
```

Needs to be converted to `TransformerLayer.forward` such that it can be found in the computed call graph. So I need to dereference the instance name to be a class name. This is something that needs type inference to determine the typer of `transformer_layer` from:

```
transformer_layer = TransformerLayer(num_heads=8, embedding_dim=embedding_dim)
```

esc replied:
I guess we could use more type annotations:

```
transformer_layer: TransformerLayer = TransformerLayer(num_heads=8, embedding_dim=embedding_dim)
```

esc replied:
> One more thing, can you add a test file in the tests directory for this chapter?
> 
> The test can just be asserting for the correct output from the function dependency of the global_calls.

I thought about this some more, and I think the best we can do for now is to test this call graph representation. The global calls will effectively be subgraphs of this adjacency list. Note that this adjacency list may contain disjoint structures.

```
########## Call Graph   ##########
{'softmax': ('np.exp', 'np.max', 'np.sum'),
 'scaled_dot_product_attention': ('query.reshape',
                                  'key.reshape',
                                  'value.reshape',
                                  'np.matmul',
                                  'np.sqrt',
                                  'softmax',
                                  'np.matmul',
                                  'context.reshape'),
 'MultiHeadAttention.__init__': (),
 'MultiHeadAttention.split_heads': ('x.reshape', 'x.transpose'),
 'MultiHeadAttention.combine_heads': ('x.transpose.reshape',),
 'MultiHeadAttention.forward': ('self.split_heads',
                                'self.split_heads',
                                'self.split_heads',
                                'scaled_dot_product_attention',
                                'self.combine_heads'),
 'FeedForwardNetwork.__init__': ('np.random.randn', 'np.random.randn'),
 'FeedForwardNetwork.forward': ('np.matmul', 'np.matmul'),
 'TransformerLayer.__init__': ('MultiHeadAttention.__init__',
                               'FeedForwardNetwork.__init__'),
 'TransformerLayer.forward': ('self.self_attn.forward',
                              'self.feed_forward.forward')}
########## ------------ ##########
```

This output is stable enough in the sense that it doesn't contain any python object IDs and is the representation that is visualized at the end of the chapter.

esc replied:
5674abca1c3d50b21dc92c43dbd7dfff4a40ae46 now has an appropriate test for ch09:

```
 💣 zsh» pytest tests/test_ch09.py
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 3.12.10, pytest-8.3.4, pluggy-1.5.0
rootdir: /Users/vhaenel/git/numba-prototypes/sealir-tutorials
plugins: xdist-3.7.0, anyio-4.7.0
collected 1 item

tests/test_ch09.py .                                                                                                                                                                                 [100%]

============================================================================================ 1 passed in 0.01s =============================================================================================
```

esc commented on sealir-tutorials/Makefile:
```diff
@@ -7,7 +7,7 @@
 
 # Define source directory and output directory
-PY_FILES := $(wildcard *.py)
+PY_FILES := $(wildcard [ch,demo]*.py)
```
👍 no problem.

esc commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -0,0 +1,326 @@
+        # Flatten the name of the call from ast.Attribute or ast.Name
+        call_qname = attribute_to_qualified_name(node)
+        class_name = self.class_stack[-1] if self.class_stack else None
+        if call_qname.startswith("self"):
```
yeah, that is nice, it's a good use of the `symtable` module, I knew it would come in handy at some point.

esc replied:
I thought about this some more and the problem I see is that, we may also end up in `classmethod` or `staticmethod` -- we may need to resolve these in future.

esc commented:
@sklam I have completed all the requested work for chapter 09 now, with the exception of the printing of the call graphs of the global calls. As discussed, this will need proper type inference and can not be easily determined from a purely static analysis of the source code.

sklam commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -321,6 +343,8 @@ def main(args):
-# TODO how to exclude this from jupyter execution.
-to_graphviz(cgv)
+if IN_NOTEBOOK:
+    to_graphviz(cgv)
```
This is not rendering, Needs a `    IPython.display(to_graphviz(cgv))`

sklam commented on sealir-tutorials/ch09_whole_program_compiler_driver.py:
```diff
@@ -123,13 +126,29 @@ def get_call_graph(self) -> dict[str: tuple[str]]:
+                            lookup(method_name).get_namespace().
+                            get_parameters()[0]
+                            if class_name and method_name else None)
+        if first_param_name and call_qname.startswith(first_param_name):
```
`call_qname.startswith(first_param_name)` can accidentally match any name with `self` prefix. Maybe split `call_qname` by `.` or track the `call_qname` as a tuple of `(objname, attrname)`

esc replied:
it is now:

```
call_qname.split(".")[0] == first_param_name
```

esc commented:
@sklam I have addressed all of your comments. On top of that I have included the improvement at:

https://github.com/numba/numba-prototypes/pull/16/commits/d6995278b3e3b5dd7fdde5c044b543c6c3f05fab

esc commented:
@sklam thank you for the merge 🎉
